### ISSUE TITLE
Bug: no-useless-default-assignment marks optional argument as non-optional when function passed to another function inline

### ISSUE THREAD

**Issue by chekrd** [typescript-eslint/typescript-eslint#11911]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=LAKAxg9gdgzgLgAgE4FcoILwIBRgEYBcOAdKQIZIDmMRaA1lBAO5QDaAugJSYB8CAbhACWAE24Y%2B%2BbJwDcoSLET5MOAA5E8ECABsApmXRY4qXeL5QU27QgD0NhAHk6CJgAtd6AF66kEBEKgRDyUIINBUKFw8TnC0bGx1BE0dfUMEYxRTXgQLK247BAAhFEoXd3QA7QDdESA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYkACAXAngA4oDG0AlobgLQrzkB2uA9AwPbWzKJLLLUATRADMAhghqi%2B5AOYMAtoibooiaNDbRIAXxDagA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEAviDUA&tokens=false

### Repro Code

```TypeScript
const run = (cb: (...args: unknown[]) => void) => cb();

run((p: boolean = true) => null)
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-useless-default-assignment": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
  }
}
```

### Expected Result

The rule does not mark this situation as useless optional parameter.

### Actual Result

The rule marks optional parameter as non optional.

**Comment by SungHyun627** [comment]
The problem seems to be in the checkAssignmentPattern function. When a parameter is matched against a rest parameter (like ...args: unknown[]) from the contextual type, it doesn't have the ts.SymbolFlags.Optional flag set. This causes the rule to incorrectly treat it as a required parameter, even though rest parameters are effectively optional.
I’d like to work on a fix for this by checking for the dotDotDotToken on the parameter’s declaration. May I take this on?

**Comment by ulrichstark** [comment]
> The problem seems to be in the checkAssignmentPattern function. When a parameter is matched against a rest parameter (like ...args: unknown[]) from the contextual type, it doesn't have the ts.SymbolFlags.Optional flag set. This causes the rule to incorrectly treat it as a required parameter, even though rest parameters are effectively optional. I’d like to work on a fix for this by checking for the dotDotDotToken on the parameter’s declaration. May I take this on?

Yes sure! Feel free to create a PR for this issue.
